launcher.py

```diff
@@ -0,0 +1,335 @@
+
+        config.save_as_yaml(launch_slurm_config["config_path_yaml"])
+        
+        # Launch the Slurm job
+        job_id = launch_slurm_job(sbatch_script)
+        print(f"🚀 Slurm job launched with id={job_id}")
+
+        # Save the Slurm script if a path is provided
+        if config.general.launch_script_path:
+            os.makedirs(config.general.launch_script_path, exist_ok=True)
+            script_filename = f"slurm_launch_script.slurm"
+            script_path = os.path.join(config.general.launch_script_path, script_filename)
+            
+            with open(script_path, 'w') as f:
+                f.write(sbatch_script)
+        print(f"    🤖 Slurm Configuration Details:")
+
+        slurm_config_keys = ['qos', 'gpus_per_node', 'cpus_per_task', 'constraint', 'account', 'reservation']
+        for key in slurm_config_keys:
+            if key in launch_slurm_config:
+                if launch_slurm_config[key] is not None:
+                    print(f"        {key}: {launch_slurm_config[key]}")
+                    
+        print("        ")
+        print("    📁 Log structure:")
+        print(f"    {config.general.logs_path}/{config.general.run}/")
+        print(f"    └── {timestamp_with_run}/")
+        print("        ├── config/")
+        print("        ├── launch-script/")
+        print("        ├── slurm-logs/")
+        if hasattr(config, 'lighteval') and config.lighteval is not None:
+            print("        └── evals/")
+            print("            ├── launch-config/")
+            print("            └── logs/")
+        else:
+            print("        └── (No evals folder)")
+
+    else:
+        # Check if running on an interactive node
+        try:
+            gpu_count = torch.cuda.device_count()
+            is_interactive = gpu_count > 0
+        except:
+            is_interactive = False
+
+        if is_interactive:
+            print("💻 Running on an interactive node with GPUs.")
+            
+            total_gpus = gpu_count
+            config_gpus = config.parallelism.dp * config.parallelism.tp * config.parallelism.pp
+            
+            if total_gpus != config_gpus:
+                raise ValueError(f"The parallelism configuration (dp={config.parallelism.dp}, tp={config.parallelism.tp}, pp={config.parallelism.pp}) "
+                                 f"doesn't match the number of available GPUs ({total_gpus}). "
+                                 f"Please adjust your configuration to match the available resources.")
+            
+            config_path_yaml = f"{config.general.config_logs_path}/launch.yaml"
+            os.makedirs("config.general.config_logs_path", exist_ok=True)
+            config.save_as_yaml(config_path_yaml)
+
+            trainer_python_file = "run_train.py"
+            cmd = f"{trainer_python_file} --config-file {args.config_path}"
+
+            launch_cmd = f"CUDA_DEVICE_MAX_CONNECTIONS='1' torchrun --nproc_per_node {gpu_count} {cmd}"
```

Comment:
There are some cases where you only want to train on 1 gpu among the 8 requested. Since it is interactive, you can just do `n_proc_per_node = min(8, config.parallelism.dp * config.parallelism.tp * config.parallelism.pp)`

Reply:
Yes, done!